### DISCUSSION TITLE
How to RTC_PIT

### DISCUSSION THREAD

**Discussion by amadeuspzs** [SpenceKonde/megaTinyCore#322] [Q&A] [answered]
I've found this [example code](https://github.com/MicrochipTech/TB3213_Getting_Started_with_RTC/blob/master/RTC_PIT_Wake_from_Sleep/main.c) from Microchip and am in the process of "porting" it.

As a relevant newcomer I have a couple of dumbo questions. If we look at:

```c
    /* Disable oscillator: */
    temp = CLKCTRL.XOSC32KCTRLA;
    temp &= ~CLKCTRL_ENABLE_bm;
    /* Writing to protected register */
    ccp_write_io((void*)&CLKCTRL.XOSC32KCTRLA, temp);
```

would this be:
```c
    /* Disable oscillator: */
    temp = CLKCTRL.XOSC32KCTRLA;
    temp &= ~CLKCTRL_ENABLE_bm;
    /* Writing to protected register */
    _PROTECTED_WRITE(CLKCTRL.XOSC32KCTRLA, temp);
```

and could it be refactored to

```c
    _PROTECTED_WRITE(CLKCTRL.XOSC32KCTRLA, ~CLKCTRL_ENABLE_bm);
```

?

**Comment by grandaspanna** [comment]
Not exactly sure what the question is, but I'm using this code to set up the PIT to trigger every 1ms:

```
void RTC_0_init()
{
  // Set up the RTC to trigger an interrupt every 1ms
  
  while (RTC.STATUS > 0)
  {
    /* Wait for all register to be synchronized */
  }
  
  RTC.CTRLA = RTC_PRESCALER_DIV1_gc   /* 1 */
    | 1 << RTC_RTCEN_bp     /* Enable: enabled */            
    | 0 << RTC_RUNSTDBY_bp; /* Run In Standby: disabled */
    
  
  // RTC.CLKSEL = RTC_CLKSEL_INT32K_gc; /* 32KHz Internal Ultra Low Power Oscillator (OSCULP32K) */
    
  while (RTC.PITSTATUS > 0) { 
    /* Wait for all register to be synchronized */
    }
    
  RTC.PITCTRLA = RTC_PERIOD_CYC32_gc  /* RTC Clock Cycles 32 */
  | 1 << RTC_PITEN_bp; /* Enable: enabled */
  
  RTC.PITINTCTRL = 1 << RTC_PI_bp; /* Periodic Interrupt: enabled */
  
}
```

This is how I'm defining the ISR for the PIT:
```
ISR(RTC_PIT_vect)
{
  /* This code is called everytime the RTC PIT interrupt is triggered */
  current_millis = millis();
  do_ya_stuff(currenr_millis);
  
  /* The interrupt flag has to be cleared manually */
  RTC.PITINTFLAGS = RTC_PI_bm;
}
```

The code was mostly generated by START, but works great in my Arduino project for a attiny412.

Side note: I was originally using TCA0 for this purpose, but found a need for PWM. The RTC seems a perfect fit for my use case.

**Reply by amadeuspzs** [reply]
Thanks for sharing! I guess I had two questions:

1. How to use the "new" AVR registers (I have been using the [ATMega328P most recently](https://github.com/amadeuspzs/atmega328p-rtc/blob/main/rtc-experiments-blink/rtc-experiments-blink.ino))
2. How to avoid using a `temp` register which feels a bit ugly?

I have got the following working well on an ATtiny1614, using the internal oscillator for testing:

```  
  uint8_t temp;  

  /* Initialize 32.768kHz Oscillator: */
  /* Disable oscillator: */
  temp = CLKCTRL.XOSC32KCTRLA;
  temp &= ~CLKCTRL_ENABLE_bm;
  
  /* Writing to protected register */
  _PROTECTED_WRITE(CLKCTRL.XOSC32KCTRLA, temp);

  while(CLKCTRL.MCLKSTATUS & CLKCTRL_XOSC32KS_bm);

  temp = CLKCTRL.XOSC32KCTRLA;
  temp &= ~CLKCTRL_SEL_bm;
  /* Writing to protected register */
  _PROTECTED_WRITE(CLKCTRL.XOSC32KCTRLA, temp);
    
  /* Enable oscillator: */
  temp = CLKCTRL.XOSC32KCTRLA;
  temp |= CLKCTRL_ENABLE_bm;
  /* Writing to protected register */
  _PROTECTED_WRITE(CLKCTRL.XOSC32KCTRLA, temp);
    
  /* Initialize RTC: */
  while (RTC.STATUS > 0);
  /* 32.768kHz External Crystal Oscillator (XOSC32K) */
  //  RTC.CLKSEL = RTC_CLKSEL_TOSC32K_gc; // external
  RTC.CLKSEL = RTC_CLKSEL_INT32K_gc; // internal
  
  RTC.PITINTCTRL = RTC_PI_bm; /* Periodic Interrupt: enabled */
  
  RTC.PITCTRLA = RTC_PERIOD_CYC32768_gc /* RTC Clock Cycles 32768 */
               | RTC_PITEN_bm; /* Enable: enabled */
```

with

```c
ISR(RTC_PIT_vect)
{
    /* Clear flag by writing '1': */
    RTC.PITINTFLAGS = RTC_PI_bm;
    
    // perform interrupt routine here
}
```

**Comment by amadeuspzs** [comment, marked as answer]
I've put together a small repo for others attempting the same: https://github.com/amadeuspzs/attiny1614-rtc

**Reply by freemovers** [reply]
Your code is missing the sleepmode and enable sleep mode, so it will never go to sleep. Also, you are setting the main clock to use an external crystal. You can accomplish the same results with the following examples: [https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/PowerSave.md](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/PowerSave.md).

**Reply by amadeuspzs** [reply]
Thanks for reviewing - I do set and enable sleep mode using the registers https://github.com/amadeuspzs/attiny1614-rtc/blob/main/attiny1614-rtc-blink/attiny1614-rtc-blink.ino#L45 - I'll move these to use the macros which is more readable.

The serial sketch doesn't yet have sleep enabled as I need to wire up an interrupt to bring it out of sleep to set the time etc.

**Reply by amadeuspzs** [reply]
In other news, I can confirm using a [currentRanger](https://lowpowerlab.com/guide/currentranger/) that power down sleep consumption is 1.2uA 👯

**Reply by freemovers** [reply]
That's great! I prefer the macro's for easier reading as well  ;)
You should be able to get even lower power consumption when you eliminate all the floating pins by defining those as an input and disable the digital input buffer as follows ([according to this application note page# 41](https://ww1.microchip.com/downloads/en/DeviceDoc/ADC-Power-Optimization-with-tinyAVR-0-1-series-and-megaAVR-0-series.pdf)):
```c
for (uint8_t pin=0; pin < 8; pin++)
{
  (&PORTA.PIN0CTRL)[pin] = PORT_ISC_INPUT_DISABLE_gc; //Disable on PAx pin
  (&PORTB.PIN0CTRL)[pin] = PORT_ISC_INPUT_DISABLE_gc; //Disable on PBx pin
  (&PORTC.PIN0CTRL)[pin] = PORT_ISC_INPUT_DISABLE_gc; //Disable on PCx pin
}
```
I have no currentranger, so I'm unable to test that, but I am curious how low you can go with the power consumption.

**Reply by amadeuspzs** [reply]
Yes [AN2515](http://ww1.microchip.com/downloads/en/AppNotes/AN2515-AVR-Low-Power-Techniques-00002515C.pdf) is good as well, and that measurement was with floating pins defined. I've remeasured the following:

Power down sleep as is: 830uA
Power down sleep with floating pins PULLUP: 1.335uA
Power down sleep with floating pins PULLUP and input disable: 1.334uA

So main gain is defining floating inputs - can't detect any notable gain with disabling input buffers.

**Reply by freemovers** [reply]
LOL, thank you for trying that. Looks like the input buffer does not make a big difference, just use the pull-up should do the trick (and makes a big difference).
